com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl

```diff
@@ -288,6 +288,17 @@ float3 ApplyShadowBias(float3 positionWS, float3 normalWS, float3 lightDirection
     return positionWS;
 }
 
+float ApplyShadowFade(float shadowAttenuation, float3 positionWS)
+{
+    float3 camToPixel = positionWS - _WorldSpaceCameraPos;
+    float distanceCamToPixel2 = dot(camToPixel, camToPixel);
+    float shadowDist = _MainLightShadowParams.z;
+    float startShadowFadeDist = shadowDist * 0.9;
+
+    float fade = saturate((distanceCamToPixel2 - startShadowFadeDist) / (shadowDist - startShadowFadeDist));
```

Comment:
only variable here is distanceCamToPixel2, this expression could be optimized to fit a single MAD instruction (Mul + Add) like following:

(distanceCamToPixel - startShadowFadeDist) / (shadowDist - startShadowFadeDist)

distanceCamToPixel / (shadowDist - startShadowFadeDist) - startShadowFadeDist / (shadowDist - startShadowFadeDist)

this becomes:
distanceCamToPixel / A - B, where A and B are constants
we can store 1/A and -B in uniform and then the expression becomes:

distanceCamToPixel2 * OneOverA + MinusB (single MAD)

We could store these in _MainLightShadowParams.z and _MainLightShadowParams.w, but it's a little bit weird these are used for additional lights as well. Do we have already another generic constant to pack those?

Reply:
I can't think of any other generic constants to pack them into which would fit better